sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/bigtable/BigtableTable.java

```diff
@@ -0,0 +1,200 @@
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.POutput;
+import org.apache.beam.sdk.values.Row;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Splitter;
+
+@Experimental
+public class BigtableTable extends SchemaBaseBeamTable implements Serializable {
+  // Should match:
+  // googleapis.com/bigtable/projects/projectId/instances/instanceId/tables/tableId"
+  private static final Pattern locationPattern =
+      Pattern.compile(
+          "(?<host>.+)/bigtable/projects/(?<projectId>.+)/instances/(?<instanceId>.+)/tables/(?<tableId>.+)");
+
+  private final String projectId;
+  private final String instanceId;
+  private final String tableId;
+  private String emulatorHost = "";
+
+  private boolean useFlatSchema = false;
+
+  private Map<String, List<String>> columnsMapping = new HashMap<>();
+
+  BigtableTable(Table table) {
+    super(table.getSchema());
+    validateSchema(schema);
+
+    String location = table.getLocation();
+    if (location == null) {
+      throw new IllegalStateException("LOCATION is required");
+    }
+    Matcher matcher = locationPattern.matcher(location);
+    validateMatcher(matcher, location);
+
+    this.projectId = getMatcherValue(matcher, "projectId");
+    this.instanceId = getMatcherValue(matcher, "instanceId");
+    this.tableId = getMatcherValue(matcher, "tableId");
+    String host = getMatcherValue(matcher, "host"); // googleapis.com or localhost:<PORT>
+    if (!"googleapis.com".equals(host)) {
+      this.emulatorHost = host;
+    }
+
+    JSONObject properties = table.getProperties();
+    if (properties.containsKey(COLUMNS_MAPPING)) {
+      columnsMapping = parseColumnsMapping(properties.getString(COLUMNS_MAPPING));
+      validateColumnsMapping(columnsMapping, schema);
+      useFlatSchema = true;
+    }
+  }
+
+  @Override
+  public PCollection<Row> buildIOReader(PBegin begin) {
+    BigtableIO.Read readTransform =
+        BigtableIO.read().withProjectId(projectId).withInstanceId(instanceId).withTableId(tableId);
+    if (!emulatorHost.isEmpty()) {
+      readTransform = readTransform.withEmulator(emulatorHost);
+    }
+    return readTransform
+        .expand(begin)
+        .apply(
+            "BigtableRowToBeamRow",
+            useFlatSchema
+                ? new BigtableRowToBeamRowFlat(schema, columnsMapping)
+                : new BigtableRowToBeamRow(schema))
+        .setRowSchema(schema);
+  }
+
+  @Override
+  public POutput buildIOWriter(PCollection<Row> input) {
+    throw new UnsupportedOperationException("Write to Cloud Bigtable is not yet supported");
+  }
+
+  @Override
+  public PCollection.IsBounded isBounded() {
+    return PCollection.IsBounded.BOUNDED;
+  }
+
+  @Override
+  public BeamTableStatistics getTableStatistics(PipelineOptions options) {
+    return BeamTableStatistics.BOUNDED_UNKNOWN;
+  }
+
+  private static Map<String, List<String>> parseColumnsMapping(String commaSeparatedMapping) {
```

Comment:
Done.

Reply:
Oh, I've remembered why they were static. That's because of null checker - it refuses to use non-static functions in the constructor (got required @Initialized and @NonNull BigtableTable)